### PR TITLE
[CPU] Add memref aliasing ops folding right before sub-type emulation

### LINKED ISSUE #14567: Failed to legalize operation 'memref.subview'
### What happened?

Open a separate issue for `memref.subview` error from #14547.

Dispatch 7 and 34 get same error
> error: failed to legalize operation 'memref.subview' that was explicitly marked illegal

`--mlir-print-ir-after-all` shows
> IR Dump After EmulateNarrowType Failed (iree-codegen-emulate-narrow-type)

https://gist.github.com/jinchen62/ef9d09696b5261562ed29e3be6cc1192
https://gist.github.com/jinchen62/962283dd39c892b212228dad3c2a2ecd

### Steps to reproduce your issue

iree-compile first_vicuna_forward_dispatch_7.mlir --iree-hal-target-backends=llvm-cpu --iree-input-type=tm_tensor --iree-vm-bytecode-module-output-format=flatbuffer-binary --iree-llvmcpu-target-cpu-features=host --iree-llvmcpu-target-triple=x86_64-linux-gnu --iree-flow-enable-data-tiling --iree-llvmcpu-enable-microkernels --iree-llvmcpu-stack-allocation-limit=256000 --iree-stream-resource-index-bits=64 --iree-vm-target-index-bits=64 --iree-vm-bytecode-module-strip-source-map=true --iree-util-zero-fill-elided-attrs --iree-vm-target-truncate-unsupported-floats --iree-codegen-check-ir-before-llvm-conversion=false

### What component(s) does this issue relate to?

_No response_

### Version information

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by hanhanW**
Fixes https://github.com/openxla/iree/issues/14567

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp [hunk 1]
```diff
@@ -131,6 +131,7 @@ static void addBufferizePasses(OpPassManager &passManager) {
   BufferizationOptions::MemCpyFn memcpyFn = cpuCopyFn;
   addIREEComprehensiveBufferizePasses(passManager, allocationFn, deallocationFn,
                                       memcpyFn);
+  passManager.addPass(memref::createFoldMemRefAliasOpsPass());
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp]: I dont think you should do this here. You should add this somewhere https://github.com/openxla/iree/blob/3926ab7cb0edd9b3ee9d0b0d7c937412aec3816f/compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp#L741 (just before you call the emulation pass)

Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp]: I thought we want the cleanup after bufferization? Those subviews are tiling artifacts, and we can remove them after bufferization. I haven't looked into why there are regressions yet..

I tried moving it before emulation pass, but there is a bug in upstream patterns. It causes a crash in some pack codegen. That's a separate issue.

**Comment by hanhanW** [comment]
It crashes if we move it to later stage. To repro: `iree-opt --fold-memref-alias-ops repro.mlir`

```mlir
func.func @matmul_accumulate_DYNxDYNxf32_times_DYNxDYNxf32_into_DYNxDYNxf32_dispatch_0_set_encoding_MATMUL_F32F32F32_LHS_DxD() {
  %c0 = arith.constant 0 : index
  %cst = arith.constant 0.000000e+00 : f32
  %c1 = arith.constant 1 : index
  %c16 = arith.constant 16 : index
  %alloca = memref.alloca() {alignment = 64 : i64} : memref<16x16xf32>
  %0 = hal.interface.constant.load[2] : i32
  %1 = hal.interface.constant.load[3] : i32
  %2 = hal.interface.constant.load[4] : i32
  %3 = hal.interface.constant.load[5] : i32
  %4 = arith.index_castui %0 : i32 to index
  %5 = arith.index_castui %1 : i32 to index
  %6 = arith.index_castui %2 : i32 to index
  %7 = arith.index_castui %3 : i32 to index
  %8 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : memref<?x?xf32>{%4, %5}
  memref.assume_alignment %8, 64 : memref<?x?xf32>
  %9 = affine.apply affine_map<()[s0] -> (s0 ceildiv 16)>()[%6]
  %10 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c0) : memref<?x?x16x1xf32>{%9, %7}
  memref.assume_alignment %10, 64 : memref<?x?x16x1xf32>
  %11 = affine.apply affine_map<()[s0] -> (s0 ceildiv 16)>()[%4]
  %workgroup_id_x = hal.interface.workgroup.id[0] : index
  %workgroup_count_x = hal.interface.workgroup.count[0] : index
  %workgroup_id_y = hal.interface.workgroup.id[1] : index
  %workgroup_count_y = hal.interface.workgroup.count[1] : index
  %12 = affine.apply affine_map<()[s0] -> (s0 * 4)>()[%workgroup_id_y]
  %13 = affine.apply affine_map<()[s0] -> (s0 * 4)>()[%workgroup_count_y]
  %14 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_id_x]
  %15 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_count_x]
  scf.for %arg0 = %12 to %11 step %13 {
    %16 = affine.min affine_map<(d0)[s0] -> (-d0 + s0 ceildiv 16, 4)>(%arg0)[%4]
    %17 = affine.min affine_map<(d0, d1)[s0] -> (d1 * -16 + s0, d0 * 16)>(%16, %arg0)[%4]
    %18 = affine.apply affine_map<(d0) -> (d0 * 16)>(%arg0)
    scf.for %arg1 = %14 to %5 step %15 {
      %19 = affine.min affine_map<(d0)[s0] -> (-d0 + s0, 64)>(%arg1)[%5]
      %subview = memref.subview %10[%arg0, %arg1, 0, 0] [%16, %19, 16, 1] [1, 1, 1, 1] : memref<?x?x16x1xf32> to memref<?x?x16x1xf32, strided<[?, 16, 1, 1], offset: ?>>
      %subview_0 = memref.subview %8[%18, %arg1] [%17, %19] [1, 1] : memref<?x?xf32> to memref<?x?xf32, strided<[?, 1], offset: ?>>
      scf.for %arg2 = %c0 to %16 step %c1 {
        %20 = affine.apply affine_map<(d0) -> (d0 * 16)>(%arg2)
        %21 = affine.min affine_map<(d0, d1) -> (16, d0 - d1 * 16)>(%17, %arg2)
        scf.for %arg3 = %c0 to %19 step %c16 {
          %22 = affine.min affine_map<(d0)[s0] -> (-d0 + s0, 16)>(%arg3)[%19]
          %23 = affine.min affine_map<(d0, d1)[s0] -> (-d1 + s0, 16, d0 - d1)>(%19, %arg3)[%19]
          %subview_1 = memref.subview %subview_0[%20, %arg3] [%21, %23] [1, 1] : memref<?x?xf32, strided<[?, 1], offset: ?>> to memref<?x?xf32, strided<[?, 1], offset: ?>>
          %subview_2 = memref.subview %subview[%arg2, %arg3, 0, 0] [1, %22, 16, 1] [1, 1, 1, 1] : memref<?x?x16x1xf32, strided<[?, 16, 1, 1], offset: ?>> to memref<1x?x16x1xf32, strided<[?, 16, 1, 1], offset: ?>>
          %24 = affine.apply affine_map<(d0, d1) -> (d0 - d1)>(%22, %23)
          %25 = arith.addi %23, %24 : index
          %subview_3 = memref.subview %alloca[0, 0] [16, %25] [1, 1] : memref<16x16xf32> to memref<16x?xf32, strided<[16, 1]>>
          scf.for %arg4 = %c0 to %c16 step %c1 {
            scf.for %arg5 = %c0 to %25 step %c1 {
              memref.store %cst, %subview_3[%arg4, %arg5] : memref<16x?xf32, strided<[16, 1]>>
            }
          }
          %subview_4 = memref.subview %subview_3[0, 0] [%21, %23] [1, 1] : memref<16x?xf32, strided<[16, 1]>> to memref<?x?xf32, strided<[16, 1]>>
          scf.for %arg4 = %c0 to %21 step %c1 {
            scf.for %arg5 = %c0 to %23 step %c1 {
              %26 = memref.load %subview_1[%arg4, %arg5] : memref<?x?xf32, strided<[?, 1], offset: ?>>
              memref.store %26, %subview_4[%arg4, %arg5] : memref<?x?xf32, strided<[16, 1]>>
            }
          }
          %expand_shape = memref.expand_shape %subview_3 [[0, 1], [2, 3]] : memref<16x?xf32, strided<[16, 1]>> into memref<1x16x?x1xf32, strided<[256, 16, 1, 1]>>
          %subview_5 = memref.subview %expand_shape[0, 0, 0, 0] [1, 16, %25, 1] [1, 1, 1, 1] : memref<1x16x?x1xf32, strided<[256, 16, 1, 1]>> to memref<1x16x?xf32, strided<[256, 16, 1]>>
          %subview_6 = memref.subview %subview_2[0, 0, 0, 0] [1, %22, 16, 1] [1, 1, 1, 1] : memref<1x?x16x1xf32, strided<[?, 16, 1, 1], offset: ?>> to memref<1x?x16xf32, strided<[?, 16, 1], offset: ?>>
          scf.for %arg4 = %c0 to %c16 step %c1 {
            scf.for %arg5 = %c0 to %25 step %c1 {
              %26 = memref.load %subview_5[%c0, %arg4, %arg5] : memref<1x16x?xf32, strided<[256, 16, 1]>>
              memref.store %26, %subview_6[%c0, %arg5, %arg4] : memref<1x?x16xf32, strided<[?, 16, 1], offset: ?>>
            }
          }
        }
      }
    }
  }
  return
}
```

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp [hunk 2]
```diff
@@ -643,6 +643,7 @@ void addCPUDataTilingPipeline(OpPassManager &passManager,
   {
     GenericVectorizationPassOptions options;
     options.vectorizePadding = true;
+    options.enableVectorMasking = true;
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp]: This change seems like it should be separate.

**Review by MaheshRavishankar — APPROVED** [review]
I think the regressions are OK for now to avoid the crash. I stamping this.

**Comment by hanhanW** [comment]
The regression is gone with enabling masking for single pack op cases. I'll send out the upstream fix, and this will be ready to land.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp (hunk 2 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp]: Yes, the change should be landed with https://github.com/openxla/iree/pull/14632

This is just for benchmarking, I will turn it ready for review after landing upstream fix.

**Review by MaheshRavishankar — APPROVED** [review]

**Comment by hanhanW** [comment]
okay, the patch I need is not in yet.. it will be in IREE in the next integrate..

**Comment by MaheshRavishankar** [comment]
> okay, the patch I need is not in yet.. it will be in IREE in the next integrate..

I will need this soon. So if it is not done by next integrate, maybe we can cherry pick. Ill pipe up when I do need it.

**Comment by hanhanW** [comment]
It's included in https://github.com/openxla/iree/pull/14678, but the integrate is not trivial.

**Comment by hanhanW** [comment]
We can cherry-pick it if it becomes a blocker.

**Pull request merged**